src/EditorFeatures/Core/Implementation/IntelliSense/AsyncCompletion/ItemManager.CompletionListUpdater.cs

```diff
@@ -0,0 +1,846 @@
+                    // typed.  i.e.  if we have "Console.$$" we don't want to commit something
+                    // like "WriteLine" since no filter text has actually been provided.  However,
+                    // if "Console.WriteL$$" is typed, then we do want "WriteLine" to be committed.
+                    for (selectedItemIndex = 0; selectedItemIndex < items.Count; ++selectedItemIndex)
+                    {
+                        if (Equals(items[selectedItemIndex].RoslynCompletionItem, bestItem))
+                            break;
+                    }
+
+                    Debug.Assert(selectedItemIndex < items.Count);
+
+                    bestOrFirstMatchResult = items[selectedItemIndex];
+                    var deduplicatedListCount = matchingItems.Count(r => !r.RoslynCompletionItem.IsPreferredItem());
+                    if (deduplicatedListCount == 1 &&
+                        _filterText.Length > 0)
+                    {
+                        uniqueItem = items[selectedItemIndex].EditorCompletionItem;
+                    }
+                }
+
+                var typedChar = _data.Trigger.Character;
+
+                // Check that it is a filter symbol. We can be called for a non-filter symbol.
+                // If inserting a non-filter character (neither IsPotentialFilterCharacter, nor Helpers.IsFilterCharacter),
+                // we should dismiss completion except cases where this is the first symbol typed for the completion session
+                // (string.IsNullOrEmpty(filterText) or string.Equals(filterText, typeChar.ToString(), StringComparison.OrdinalIgnoreCase)).
+                // In the latter case, we should keep the completion because it was confirmed just before in InitializeCompletion.
+                if (FilterReason == CompletionFilterReason.Insertion &&
+                    !string.IsNullOrEmpty(_filterText) &&
+                    !string.Equals(_filterText, typedChar.ToString(), StringComparison.OrdinalIgnoreCase) &&
+                    !IsPotentialFilterCharacter(typedChar) &&
+                    !Helpers.IsFilterCharacter(bestOrFirstMatchResult.RoslynCompletionItem, typedChar, _filterText))
+                {
+                    return null;
+                }
+
+                var isHardSelection = IsHardSelection(
+                    _filterText, bestOrFirstMatchResult.RoslynCompletionItem, bestOrFirstMatchResult.MatchedFilterText, _hasSuggestedItemOptions);
+
+                var updateSelectionHint = isHardSelection ? UpdateSelectionHint.Selected : UpdateSelectionHint.SoftSelected;
+
+                return new(selectedItemIndex, updateSelectionHint, uniqueItem);
+            }
+
+            private ItemSelection? HandleDeletionTrigger(IReadOnlyList<MatchResult<VSCompletionItem>> items)
+            {
+                var matchingItems = items.Where(r => r.MatchedFilterText);
+                if (TriggerReason == CompletionTriggerReason.Insertion &&
```

Comment:
i'm confused.  if this is HandleDeletionTrigger, how can the trigger reason be 'insertion'?

Reply:
Deletion is the initial trigger reason (i.e. the action triggered this completion session), and insertion is what caused the completion list to be updated. When combined, it means you first hit "backspace" which triggers the completion, and then you continue typing so the list keep refreshing.